tests/nexus/test_handler.py

```diff
@@ -0,0 +1,901 @@
+    # timeouts on executions they start in their handler.
+    operation = "check_operation_timeout_header"
+    headers = {"Operation-Timeout": "10ms"}
+    expected = SuccessfulResponse(
+        status_code=200,
+    )
+
+
+class BadRequest(_FailureTestCase):
+    operation = "echo"
+    input = Input(7)  # type: ignore
+    expected = UnsuccessfulResponse(
+        status_code=400,
+        retryable_header=False,
+        failure_message=lambda s: s.startswith("Failed converting field"),
+    )
+
+
+class NonRetryableApplicationError(_FailureTestCase):
+    operation = "non_retryable_application_error"
+    expected = UnsuccessfulResponse(
+        status_code=500,
+        retryable_header=False,
+        retryable_exception=False,
+        failure_message="non-retryable application error",
+    )
+
+    @classmethod
+    def check_response(
+        cls, response: httpx.Response, with_service_definition: bool
+    ) -> None:
+        super().check_response(response, with_service_definition)
+        failure = Failure(**response.json())
+        err = failure.exception
+        assert isinstance(err, ApplicationError)
+        assert err.non_retryable
+        assert err.type == "TestFailureType"
+        assert err.details == ("details arg",)
+
+
+class RetryableApplicationError(_FailureTestCase):
+    operation = "retryable_application_error"
+    expected = UnsuccessfulResponse(
+        status_code=500,
+        retryable_header=True,
+        failure_message="retryable application error",
+    )
+
+
+class HandlerErrorInternal(_FailureTestCase):
+    operation = "handler_error_internal"
+    expected = UnsuccessfulResponse(
+        status_code=500,
+        # TODO(nexus-prerelease): check this assertion
+        retryable_header=False,
+        failure_message="cause message",
+    )
+
+
+class OperationError(_FailureTestCase):
+    operation = "operation_error_failed"
+    expected = UnsuccessfulResponse(
+        status_code=424,
+        # TODO(nexus-prerelease): check that OperationError should not set retryable header
+        retryable_header=None,
+        failure_message="deliberate operation error",
+        headers=UNSUCCESSFUL_RESPONSE_HEADERS | {"nexus-operation-state": "failed"},
+    )
+
+
+class UnknownService(_FailureTestCase):
+    operation = "echo"
+    service = "NonExistentService"
+    expected = UnsuccessfulResponse(
+        status_code=404,
+        retryable_header=False,
+        failure_message="No handler for service 'NonExistentService'.",
+    )
+
+
+class UnknownOperation(_FailureTestCase):
+    operation = "NonExistentOperation"
+    expected = UnsuccessfulResponse(
+        status_code=404,
+        retryable_header=False,
+        failure_message=lambda s: s.startswith(
+            "Nexus service contract 'MyService' has no operation 'NonExistentOperation'."
+        ),
+    )
+
+
+@pytest.mark.parametrize(
```

Comment:
IMO these large parameterized test cases are hard to read and reason about (have to jump around quite a bit to see the whole picture). If it's not too much extra code, can we have separate tests that maybe use limited helpers where needed? We've had flakes and maintenance needs on tests like `test_unfinished_handler_on_workflow_termination` and such and they are a bit rough on developers coming to maintain.

Reply:
No, this is solid usage of `@pytest.mark.parametrize`. I agree with you that tests can become very hard to read when it's used wrongly (and I've been guilty of that in my workflow caller tests which need refactoring along the lines you describe). But the usage here is good style and appropriate: Each entry in the parameterized list is a test case; the test itself just makes generic assertions; the way one reads the test code is to read the test cases.

> We've had flakes and maintenance needs on tests like `test_unfinished_handler_on_workflow_termination`

That's true. They appear to be timing issues; I'm not sure we have reason to blame parametrized there, but I may be wrong. In any case, not relevant here.